k8s.go

```diff
@@ -0,0 +1,643 @@
+//    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//    License for the specific language governing permissions and limitations
+//    under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	backendapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/compat"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s/resources"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+
+	svbgpconfig "github.com/osrg/gobgp/config"
+	svbgptable "github.com/osrg/gobgp/table"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	kapiv1 "k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	Act_add  = "add"
+	Act_upd  = "upd"
+	Act_del  = "del"
+	Act_same = "same"
+)
+
+type ActionList struct {
+	Add  []string
+	Upd  []string
+	Del  []string
+	Same []string
+}
+
+func CompareMap(lasts map[string]string, currs map[string]string) ActionList {
+	act := ActionList{}
+	for key, last := range lasts {
+		if cur, ok := currs[key]; ok == false {
+			act.Del = append(act.Del, key)
+		} else if last != cur {
+			act.Upd = append(act.Upd, key)
+		} else {
+			act.Same = append(act.Same, key)
+		}
+	}
+	for key, _ := range currs {
+		if _, ok := lasts[key]; ok == false {
+			act.Add = append(act.Add, key)
+		}
+	}
+	return act
+}
+
+// populateFromKVPairs populates the vars KV map from the supplied set of
+// KVPairs.  This uses the libcalico-go compat module and serialization functions
+// to write out the KVPairs in etcdv2 format.  This works in conjunction with the
+// etcdVarClient defined below which provides a "mock" etcd backend which actually
+// just writes out data to the vars map.
+func populateFromKVPairs(kvps []*model.KVPair, vars map[string]string) {
+	// Create a etcdVarClient to write the KVP results in the vars map, using the
+	// compat adaptor to write the values in etcdv2 format.
+	client := compat.NewAdaptor(&etcdVarClient{vars: vars})
+	for _, kvp := range kvps {
+		if _, err := client.Apply(kvp); err != nil {
+			log.Error("Failed to convert k8s data to etcdv2 equivalent: %s = %s", kvp.Key, kvp.Value)
+		}
+	}
+}
+
+type IntervalProcessor struct {
+	interval int
+	k8scli   *k8sClient
+	ipam     *ipamCacheK8s
+}
+
+func (p *IntervalProcessor) IntervalLoop() error {
+	if err := p.k8scli.updatePrefix(); err != nil {
+		return err
+	}
+	if err := p.k8scli.initialNeighborConfigSetting(); err != nil {
+		return err
+	}
+	ippool, err := p.ipam.getIPPools()
+	if err != nil {
+		return err
+	}
+	p.ipam.lastIPPool = ippool
+	for {
+		log.Debug("polling")
+		p.ipam.sync()
+		p.k8scli.checkBGPConfig()
+		select {
```

Comment:
time.Sleep would be enough, IMO

Reply:
I thought some channels may be used.

Reply:
btw, even with the channels, you don't need a `select` if it's only waiting for one channel. It's a blocking call so `<-time.After...` will block. But I agree with @toshiiw `time.Sleep` is more readable

Reply:
I thought that in the future we may add some channels. But  I will use it as using time.Sleep is simpler in the current codes.

Reply:
Done.